File: 03-java-socket/src/main/java/org/protelis/demo/ProtelisConfigSpec.java
package org.protelis.demo;

import com.uchuhimo.konf.ConfigSpec;
import com.uchuhimo.konf.RequiredItem;
import org.protelis.demo.data.ProtelisNode;

import java.util.List;

/**
 * Konf specification for the TOML configuration file.
 */
public final class ProtelisConfigSpec {

    /**
     * Create a new config specification named protelis.
     */
    public static final ConfigSpec SPEC = new ConfigSpec("protelis");
    /**
     * Required field iterations. It represents the number of cycles the simulation will perform.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<Integer> iterations = new RequiredItem<>(SPEC, "iterations") { };
    /**
     * Required field protelisModuleName. It is the name of the .pt source file.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<String> protelisModuleName = new RequiredItem<>(SPEC, "protelisModuleName") { };
    /**
     * Required field nodes. It contains a list of protelis nodes.
     */
    @SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<List<ProtelisNode>> nodes = new RequiredItem<>(SPEC, "nodes") { };

    private ProtelisConfigSpec() { }
}

File: 04-kotlin-socket/src/main/kotlin/org/protelis/demo/ProtelisConfigSpec.kt
package org.protelis.demo

import com.uchuhimo.konf.ConfigSpec

/**
 * An IPv4 [host] with its [port].
 */
data class IPv4Host(
    val host: String,
    val port: Int,
)

/**
 * A networked Protelis node with IPv4 [hostandport], an [id], and a predefined set of [neighbors].
 * It can possibly be a [leader] (defaults to `false`)
 */
data class ProtelisNode(
    val hostandport: IPv4Host,
    val id: Int,
    val neighbors: Set<IPv4Host>,
    val leader: Boolean = false,
)

/**
 * Demo configuration.
 */
object ProtelisConfigSpec : ConfigSpec("protelis") {
    /**
     * Iterations to be performed.
     */
    val iterations by required<Int>(description = "Number of iterations the devices will run.")

    /**
     * Protelis program to be executed.
     */
    val protelisModuleName by required<String>(description = "Protelis program to be executed")

    /**
     * Nodes participating in the system.
     */
    val nodes by required<List<ProtelisNode>>(description = "List of nodes")
}
